### PR TITLE
Scheduler refactor Pt. 1

### PR DESCRIPTION
**Pull request by kevinhartman**
* Simplifies scheduling logic, specifically regarding thread status. It should be much clearer which statuses are valid for a thread at any given point in the system.
* Removes some dead code from thread.cpp.
* Moves the implementation of resetting a ThreadContext to the corresponding core's implementation.

I have a more aggressive refactor in progress as well (not yet passing all waitsync tests) which builds on top of these changes.

### CODE REVIEW

File: src/core/hle/kernel/thread.cpp [hunk 1]
```diff
@@ -38,87 +41,62 @@ static Common::ThreadQueueList<Thread*, THREADPRIO_LOWEST+1> thread_ready_queue;
 
 static Thread* current_thread;
 
-static const u32 INITIAL_THREAD_ID = 1; ///< The first available thread id at startup
-static u32 next_thread_id; ///< The next available thread id
+inline static u32 const NewThreadId() {
+    static u32 thread_id = 1; // The first available thread id at startup
```
Comment (yuriks) [inline src/core/hle/kernel/thread.cpp]: This variable should be kept outside the function, since it's impossible to savestate it like this.

File: src/core/hle/kernel/thread.cpp [hunk 2]
```diff
@@ -38,87 +41,62 @@ static Common::ThreadQueueList<Thread*, THREADPRIO_LOWEST+1> thread_ready_queue;
...
+    return THREADSTATUS_WAIT_ARB == thread->status && wait_address == thread->wait_address;
 }
 
-/// Stops the current thread
-void Thread::Stop(const char* reason) {
-    // Release all the mutexes that this thread holds
-    ReleaseThreadMutexes(this);
+/// Exits the current thread
+void ExitCurrentThread() {
+    Thread* thread = GetCurrentThread();
```
Comment (yuriks) [inline src/core/hle/kernel/thread.cpp]: There's no reason for this to implicitly take the current thread.

File: src/core/hle/kernel/thread.cpp [hunk 3]
```diff
@@ -151,84 +129,78 @@ Thread* ArbitrateHighestPriorityThread(u32 address) {
...
     thread->wait_set_output = wait_set_output;
     thread->wait_all = wait_all;
-
-    // It's possible to call WaitSynchronizationN without any objects passed in...
-    if (wait_object != nullptr)
-        thread->wait_objects.push_back(wait_object);
-
-    ChangeThreadState(thread, ThreadStatus(THREADSTATUS_WAIT | (thread->status & THREADSTATUS_SUSPEND)));
+    thread->wait_objects.clear();
+    thread->wait_objects.insert(thread->wait_objects.end(), wait_objects.begin(), wait_objects.end());
```
Comment (yuriks) [inline src/core/hle/kernel/thread.cpp]: You can use vector::assign to do this in one step. Or even better, use vector::swap.

File: src/core/hle/kernel/thread.cpp (hunk 2 above)
Comment (kevinhartman) [inline src/core/hle/kernel/thread.cpp]: The reason I did this is because it's not valid for a thread that's not the current thread to be exited, just as it's not valid for Wait* functions to operate on arbitrary threads.

Comment (yuriks) [inline src/core/hle/kernel/thread.cpp]: > it's not valid for a thread that's not the current thread to be exited
Why not? This is required for the SVC TerminateProcess.

File: src/core/hle/kernel/thread.cpp (hunk 3 above)
Comment (kevinhartman) [inline src/core/hle/kernel/thread.cpp]: Thought about using vector::swap, but wasn't sure the caller would expect that behavior. I'll look into alternatives.

File: src/core/hle/kernel/thread.cpp (hunk 1 above)
Comment (kevinhartman) [inline src/core/hle/kernel/thread.cpp]: Ah hadn't thought of that. I'll move it out.

File: src/core/hle/kernel/thread.cpp (hunk 3 above)
Comment (yuriks) [inline src/core/hle/kernel/thread.cpp]: You're taking the vector by value in this function, there's nothing for the caller to expect. But I just realized that an assignment with move will accomplish the same thing, and will be less weird than swap.

Comment (kevinhartman) [inline src/core/hle/kernel/thread.cpp]: Oh you're right, cool! I'll use move. I was thinking about it like an array rather than an object with metadata and such when I hastily dismissed swap.

File: src/core/hle/kernel/thread.cpp (hunk 2 above)
Comment (kevinhartman) [inline src/core/hle/kernel/thread.cpp]: Good point! I'll add the parameter back.

File: src/core/hle/svc.cpp [hunk 4]
```diff
@@ -201,12 +200,15 @@ static Result WaitSynchronizationN(s32* out, Handle* handles, s32 handle_count,
...
         // Actually wait the current thread on each object if we decided to wait...
+        std::vector<SharedPtr<Kernel::WaitObject>> wait_objects;
         for (int i = 0; i < handle_count; ++i) {
             auto object = Kernel::g_handle_table.GetWaitObject(handles[i]);
             object->AddWaitingThread(Kernel::GetCurrentThread());
-            Kernel::WaitCurrentThread_WaitSynchronization(object, true, wait_all);
+            wait_objects.push_back(object);
         }
 
+        Kernel::WaitCurrentThread_WaitSynchronization(wait_objects, true, wait_all);
```
Comment (yuriks) [inline src/core/hle/svc.cpp]: Use `std::move` here to move the vector into the function.

File: src/core/arm/interpreter/arm_interpreter.cpp [hunk 5]
```diff
@@ -77,6 +77,15 @@ void ARM_Interpreter::ExecuteInstructions(int num_instructions) {
     ARMul_Emulate32(state);
 }
 
+void ARM_Interpreter::ResetContext(Core::ThreadContext& context, u32 stack_top, u32 entry_point, u32 arg) {
```
Comment (Subv) [inline src/core/arm/interpreter/arm_interpreter.cpp]: You can remove this, armemu is no more!

File: src/core/hle/kernel/thread.cpp [hunk 6]
```diff
@@ -38,87 +41,61 @@ static Common::ThreadQueueList<Thread*, THREADPRIO_LOWEST+1> thread_ready_queue;
...
+    thread->status = reason;
 }
 
 /// Check if a thread is waiting on a the specified wait object
 static bool CheckWait_WaitObject(const Thread* thread, WaitObject* wait_object) {
-    auto itr = std::find(thread->wait_objects.begin(), thread->wait_objects.end(), wait_object);
-
-    if (itr != thread->wait_objects.end())
-        return thread->IsWaiting();
+    if (THREADSTATUS_WAIT_SYNC != thread->status) return false;
```
Comment (yuriks) [inline src/core/hle/kernel/thread.cpp]: Split this line.

File: src/core/hle/svc.cpp [hunk 7]
```diff
@@ -209,12 +208,15 @@ static ResultCode WaitSynchronizationN(s32* out, Handle* handles, s32 handle_cou
     if (wait_thread) {
 
         // Actually wait the current thread on each object if we decided to wait...
+        std::vector<SharedPtr<Kernel::WaitObject>> wait_objects;
```
Comment (yuriks) [inline src/core/hle/svc.cpp]: do `wait_objects.reserve` here to pre-allocate memory.

File: src/core/hle/kernel/thread.h [hunk 8]
```diff
@@ -158,6 +151,11 @@ void WaitCurrentThread_ArbitrateAddress(VAddr wait_address);
 void WakeThreadAfterDelay(Thread* thread, s64 nanoseconds);
 
 /**
+ * Exits the current thread, invalidating it from further use.
+ */
+void ExitThread(Thread* thread);
```
Comment (yuriks) [inline src/core/hle/kernel/thread.h]: This can be a member of Thread like it used to.

File: src/core/hle/kernel/thread.cpp [hunk 9]
```diff
@@ -47,83 +52,59 @@ Thread* GetCurrentThread() {
...
-    status = THREADSTATUS_DORMANT;
-    WakeupAllWaitingThreads();
+    // Cancel any outstanding wakeup events for this thread
+    CoreTiming::UnscheduleEvent(ThreadWakeupEventType, callback_handle);
 
+    MakeNotReady(this, THREADSTATUS_DEAD);
+    
+    WakeupAllWaitingThreads();
+	
+	// TODO(peachum): is this needed? Also, should check if it's waiting on sync at all before looking at wait_objects
```
Comment (kevinhartman) [inline src/core/hle/kernel/thread.cpp]: Forgot to address this. I'll do that and add a commit.

**Comment by kevinhartman** [comment]
@bunnei This is rebased and ready for another look, by the way ;)

File: src/core/arm/arm_interface.h [hunk 10]
```diff
@@ -86,6 +86,11 @@ class ARM_Interface : NonCopyable {
     virtual void AddTicks(u64 ticks) = 0;
 
     /**
+     * Initializes a CPU context for use on this CPU
+     */
```
Comment (bunnei) [inline src/core/arm/arm_interface.h]: Please add the parameters to the docstring

File: src/core/arm/dyncom/arm_dyncom.cpp [hunk 11]
```diff
@@ -93,6 +93,16 @@ void ARM_DynCom::ExecuteInstructions(int num_instructions) {
...
 }
 
+void ARM_DynCom::ResetContext(Core::ThreadContext& context, u32 stack_top, u32 entry_point, u32 arg) {
+    memset(&context, 0, sizeof(Core::ThreadContext));
+
+    context.cpu_registers[0] = arg;
+    context.pc = entry_point;
+    context.sp = stack_top;
+    context.cpsr = 0x1F; // Usermode
+    context.mode = 8;    // Instructs dyncom CPU core to start execution as if it's "resuming" a thread.
```
Comment (bunnei) [inline src/core/arm/dyncom/arm_dyncom.cpp]: Are we sure this hack is still necessary? I thought this was just an armemu thing...

File: src/core/arm/dyncom/arm_dyncom.h [hunk 12]
```diff
@@ -68,6 +68,11 @@ class ARM_DynCom final : virtual public ARM_Interface {
     void AddTicks(u64 ticks) override;
 
     /**
+     * Initializes a CPU context for use on this CPU
+     */
```
Comment (bunnei) [inline src/core/arm/dyncom/arm_dyncom.h]: No need for duplicate docstring (it's defined already in the parent class)

File: src/core/hle/kernel/thread.cpp [hunk 13]
```diff
@@ -37,8 +40,10 @@ static Common::ThreadQueueList<Thread*, THREADPRIO_LOWEST+1> thread_ready_queue;
 
 static Thread* current_thread;
 
-static const u32 INITIAL_THREAD_ID = 1; ///< The first available thread id at startup
-static u32 next_thread_id; ///< The next available thread id
+static u32 next_thread_id = 1; // The first available thread id at startup
```
Comment (bunnei) [inline src/core/hle/kernel/thread.cpp]: Why did you remove the doxy-style comment?

File: src/core/hle/kernel/thread.cpp [hunk 14]
```diff
@@ -37,8 +40,10 @@ static Common::ThreadQueueList<Thread*, THREADPRIO_LOWEST+1> thread_ready_queue;
...
+inline static u32 const NewThreadId() {
```
Comment (bunnei) [inline src/core/hle/kernel/thread.cpp]: Please add a function comment

File: src/core/hle/kernel/thread.cpp [hunk 15]
```diff
@@ -47,83 +52,58 @@ Thread* GetCurrentThread() {
...
-    t->context.cpsr = 0x1F; // Usermode
-
-    // TODO(bunnei): This instructs the CPU core to start the execution as if it is "resuming" a
-    // thread. This is somewhat Sky-Eye specific, and should be re-architected in the future to be
-    // agnostic of the CPU core.
-    t->context.mode = 8;
-
-    if (t->current_priority < lowest_priority) {
-        t->current_priority = t->initial_priority;
+/// Makes a thread available for scheduling
```
Comment (bunnei) [inline src/core/hle/kernel/thread.cpp]: Might as well add a full doxygen comment here as well

File: src/core/hle/kernel/thread.cpp [hunk 16]
```diff
@@ -47,83 +52,58 @@ Thread* GetCurrentThread() {
...
+static void MakeReady(Thread* thread) {
```
Comment (bunnei) [inline src/core/hle/kernel/thread.cpp]: Not sure that I like the `MakeReady` and `MakeNotReady` naming

File: src/core/hle/kernel/thread.cpp [hunk 17]
```diff
@@ -47,83 +52,58 @@ Thread* GetCurrentThread() {
...
-            thread_ready_queue.remove(t->current_priority, t);
-        }
-    }  else if (ready) {
-        if (t->IsRunning()) {
-            thread_ready_queue.push_front(t->current_priority, t);
-        } else {
-            thread_ready_queue.push_back(t->current_priority, t);
-        }
-        t->status = THREADSTATUS_READY;
+static void MakeNotReady(Thread* thread, ThreadStatus reason) {
```
Comment (bunnei) [inline src/core/hle/kernel/thread.cpp]: As above, please add a doxygen comment

File: src/core/arm/dyncom/arm_dyncom.cpp (hunk 11 above)
Comment (kevinhartman) [inline src/core/arm/dyncom/arm_dyncom.cpp]: Not sure at all. Just moved this from thread.cpp, and the comment made me assume it was necessary for dyncom only.

File: src/core/hle/kernel/thread.cpp [hunk 18]
```diff
@@ -156,83 +136,77 @@ Thread* ArbitrateHighestPriorityThread(u32 address) {
...
-static void CallThread(Thread* t) {
-    // Stop waiting
-    ChangeThreadState(t, THREADSTATUS_READY);
-}
-
 /// Switches CPU context to that of the specified thread
-static void SwitchContext(Thread* t) {
-    Thread* cur = GetCurrentThread();
+static void SwitchContext(Thread* thread) {
+
```
Comment (bunnei) [inline src/core/hle/kernel/thread.cpp]: Remove extra line

File: src/core/arm/dyncom/arm_dyncom.h (hunk 12 above)
Comment (kevinhartman) [inline src/core/arm/dyncom/arm_dyncom.h]: Should I remove the docstrings from the other definitions in this interface as well? I'm definitely on board with only have documentation once on the base's definition.

File: src/core/hle/kernel/thread.cpp [hunk 19]
```diff
@@ -156,83 +136,77 @@ Thread* ArbitrateHighestPriorityThread(u32 address) {
...
 
     // Save context for current thread
-    if (cur) {
-        Core::g_app_core->SaveContext(cur->context);
+    if (last) {
+        Core::g_app_core->SaveContext(last->context);
 
-        if (cur->IsRunning()) {
-            ChangeReadyState(cur, true);
+        if (THREADSTATUS_RUNNING == last->status) {
```
Comment (bunnei) [inline src/core/hle/kernel/thread.cpp]: Fwiw while I initially pushed for this ordering with if-statements, the rest of the team veto'd it a while ago. So the preferred style would be `if (last->status == THREADSTATUS_RUNNING)`

File: src/core/hle/kernel/thread.h [hunk 20]
```diff
@@ -137,7 +135,7 @@ class Thread final : public WaitObject {
 extern SharedPtr<Thread> g_main_thread;
 
 /// Sets up the primary application thread
```
Comment (bunnei) [inline src/core/hle/kernel/thread.h]: Might as well add a more complete docstring here as well

File: src/core/hle/kernel/thread.h [hunk 21]
```diff
@@ -99,6 +92,11 @@ class Thread final : public WaitObject {
      */
     void SetWaitSynchronizationOutput(s32 output);
 
+    /**
+     * Stops a thread, invalidating it from further use.
+     */
+    void Stop();
```
Comment (bunnei) [inline src/core/hle/kernel/thread.h]: Docstrings should just use `///` if its just one line

File: src/core/hle/kernel/thread.h [hunk 22]
```diff
@@ -31,13 +31,13 @@ enum ThreadProcessorId {
...
-    THREADSTATUS_DORMANT        = 16,
-    THREADSTATUS_DEAD           = 32,
-    THREADSTATUS_WAITSUSPEND    = THREADSTATUS_WAIT | THREADSTATUS_SUSPEND
+    THREADSTATUS_RUNNING,
+    THREADSTATUS_READY,
+    THREADSTATUS_WAIT_ARB,
+    THREADSTATUS_WAIT_SLEEP,
+    THREADSTATUS_WAIT_SYNC,
+    THREADSTATUS_DORMANT,
+    THREADSTATUS_DEAD
```
Comment (bunnei) [inline src/core/hle/kernel/thread.h]: What's the difference between dormant and dead at this point?

File: src/core/hle/kernel/thread.cpp (hunk 13 above)
Comment (kevinhartman) [inline src/core/hle/kernel/thread.cpp]: Without having used doxy, I assumed that it shouldn't be doxy-documented because it's not visible in any interface. Should I use that style anyway for consistency?

File: src/core/hle/kernel/thread.cpp (hunk 16 above)
Comment (kevinhartman) [inline src/core/hle/kernel/thread.cpp]: Yeah me either. In some further changes, I've gotten rid of both of these. In the meantime, I can try to rename perhaps.

File: src/core/hle/kernel/thread.h (hunk 21 above)
Comment (kevinhartman) [inline src/core/hle/kernel/thread.h]: Aw man but what about the prettiness? :(

File: src/core/hle/kernel/thread.h (hunk 22 above)
Comment (kevinhartman) [inline src/core/hle/kernel/thread.h]: Dormant means that the thread has been created, but not scheduled. Dead means that it's exited and cannot be used. In the current codebase, CreateThread immediately schedules a created thread (making that useless). I way toying with the idea of having CreateThread return a dormant Thread, allowing the caller to then schedule the thread on some implementation of a scheduler later on, but it's looking like there wouldn't be a need for that (SVC CreateThread schedules threads immediately (and switched to them)).

**Comment by kevinhartman** [comment]
@bunnei Comments have been addressed. Also did some more refactoring wrt. thread status changes.

File: src/core/hle/kernel/thread.h [hunk 23]
```diff
@@ -55,22 +65,29 @@ class Thread final : public WaitObject {
...
-    inline bool IsReady() const { return (status & THREADSTATUS_READY) != 0; }
-    inline bool IsWaiting() const { return (status & THREADSTATUS_WAIT) != 0; }
-    inline bool IsSuspended() const { return (status & THREADSTATUS_SUSPEND) != 0; }
-    inline bool IsIdle() const { return idle; }
-
     bool ShouldWait() override;
     void Acquire() override;
 
+    /**
+     * Returns true if the thread is the idle (stub) thread. False otherwise.
```
Comment (bunnei) [inline src/core/hle/kernel/thread.h]: I know it sounds pedantic, but for the sake of nice generated doygen documentation, an @return should be here

File: src/core/hle/kernel/thread.h [hunk 24]
```diff
@@ -55,22 +65,29 @@ class Thread final : public WaitObject {
...
+     */
+    inline bool IsIdle() const { return idle; }
+
+    /**
+     * Gets the thread's current priority
+     */
```
Comment (bunnei) [inline src/core/hle/kernel/thread.h]: ditto

File: src/core/hle/kernel/thread.h [hunk 25]
```diff
@@ -55,22 +65,29 @@ class Thread final : public WaitObject {
...
+
+    /**
+     * Sets the thread's current priority
+     * @param priority The new priority
+     */
     void SetPriority(s32 priority);
 
+    /**
+     * Gets the thread's thread ID
+     */
```
Comment (bunnei) [inline src/core/hle/kernel/thread.h]: ditto

File: src/core/hle/kernel/thread.h [hunk 26]
```diff
@@ -31,13 +31,13 @@ enum ThreadProcessorId {
...
-    THREADSTATUS_WAIT           = 4,
-    THREADSTATUS_SUSPEND        = 8,
-    THREADSTATUS_DORMANT        = 16,
-    THREADSTATUS_DEAD           = 32,
-    THREADSTATUS_WAITSUSPEND    = THREADSTATUS_WAIT | THREADSTATUS_SUSPEND
+    THREADSTATUS_RUNNING,
+    THREADSTATUS_READY,
+    THREADSTATUS_WAIT_ARB,
+    THREADSTATUS_WAIT_SLEEP,
+    THREADSTATUS_WAIT_SYNC,
```
Comment (bunnei) [inline src/core/hle/kernel/thread.h]: I'd prefer this be called `THREADSTATUS_WAIT_SYNCH`, but perhaps that just preference (I feel like we abbreviate "synchronization" as "synch" elsewhere)

**Comment by bunnei** [comment]
Added a few very minor comments, but LGTM. I'd like @yuriks feedback before merging, and also a rebase.

**Comment by kevinhartman** [comment]
Comments addressed. Waiting on @yuriks and then I'll rebase.

File: src/core/arm/dyncom/arm_dyncom.cpp (hunk 11 above)
Comment (yuriks) [inline src/core/arm/dyncom/arm_dyncom.cpp]: I looked at a bit and I'm fairly confident this can be removed entirely:
 - `mode` itself is only used when saving/restoring to/from `ARMul_State::NextInstr`.
 - `NextInstr` is (apart from the saving to mode) never read from, so it's also dead code.

File: src/core/arm/dyncom/arm_dyncom.h [hunk 27]
```diff
@@ -17,75 +17,30 @@ class ARM_DynCom final : virtual public ARM_Interface {
     ARM_DynCom();
     ~ARM_DynCom();
 
-    /**
-     * Set the Program Counter to an address
-     * @param pc Address to set PC to
-     */
     void SetPC(u32 pc) override;
```
Comment (yuriks) [inline src/core/arm/dyncom/arm_dyncom.h]: Can you remove all these blank lines? They don't really serve any purpose. Perhaps leave them where they divide sections between Get/Sets, Ticks and Context functions.

File: src/core/hle/kernel/thread.cpp [hunk 28]
```diff
@@ -385,75 +375,76 @@ ResultVal<SharedPtr<Thread>> Thread::Create(std::string name, VAddr entry_point,
...
     // TODO(yuriks): Propagate error
     _dbg_assert_(Kernel, thread_res.Succeeded());
-    SharedPtr<Thread> thread = std::move(*thread_res);
 
-    // If running another thread already, set it to "ready" state
-    Thread* cur = GetCurrentThread();
-    if (cur && cur->IsRunning()) {
-        ChangeReadyState(cur, true);
-    }
+    SharedPtr<Thread> thread = std::move(*thread_res);
```
Comment (yuriks) [inline src/core/hle/kernel/thread.cpp]: the assert and `std::move(*thread_res)` can be replaced with `thread_res.MoveFrom()`.

File: src/core/hle/kernel/thread.h [hunk 29]
```diff
@@ -31,13 +31,13 @@ enum ThreadProcessorId {
...
-    THREADSTATUS_DORMANT        = 16,
-    THREADSTATUS_DEAD           = 32,
-    THREADSTATUS_WAITSUSPEND    = THREADSTATUS_WAIT | THREADSTATUS_SUSPEND
+    THREADSTATUS_RUNNING,
+    THREADSTATUS_READY,
+    THREADSTATUS_WAIT_ARB,
+    THREADSTATUS_WAIT_SLEEP,
+    THREADSTATUS_WAIT_SYNCH,
+    THREADSTATUS_DORMANT,
+    THREADSTATUS_DEAD
```
Comment (yuriks) [inline src/core/hle/kernel/thread.h]: Please add documentation explaining what each status represents and in what situations a thread is in it.

**Comment by yuriks** [comment]
A few more comments, after those it'll LGTM

**Comment by kevinhartman** [comment]
@bunnei @yuriks All set!

**Comment by bunnei** [comment]
Proposed new ThreadStatus docstrings: https://gist.github.com/bunnei/117f1ef824af68e829cb
Then I'll merge!
